proposals/mainnet_migration.md

```diff
@@ -0,0 +1,528 @@
+
+### Assigning Generator Key to Validator Accounts
+
+In the protocol used by Lisk Core v4, each validator needs a [generator key](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0044.md#generator-key) in order to generate blocks. Moreover, the DPoS module expects that each validator account has such a key specified in the snapshot block. Otherwise, the snapshot block is invalid (see [here](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#genesis-state-initialization)). Therefore, it is aimed to use the existing public keys of the validator accounts as the initial generator keys in the snapshot block. However, the public key does not belong to the [account state](https://lisk.com/documentation/lisk-sdk/protocol/accounts.html) anymore with Lisk Core v3\. For this reason, the whole block history from the previous snapshot block onwards is scanned for public keys of registered validators. The found keys are used as generator keys in the snapshot block. Validator accounts for which no public key was found get an "invalid" public key assigned, i.e., one for which every signature validation will fail. The specific value is discussed in the [following subsection](#invalid-ed25519-public-key). Those validators should send an [update generator key](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#update-generator-key) transaction as soon as possible once Lisk Core v4 is used, ideally within the [bootstrap period](#bootstrap-period). Otherwise, those validators can still be active validators, but are not able to generate valid blocks as they cannot create [valid block signatures](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0055.md#signature). In the worst case, this may lead to banning which happens after 50 blocks in a row are missed.
+
+To avoid the need to send an _updated generator key_ right after the hard fork, registered validators that did not forge a block and did not send a transaction in the Lisk Core v3 history could simply send a transaction any time before the hard fork. Then the sender public key of this transaction will be set as the generator key as described above. Nevertheless, every validator is encouraged to perform an _update generator key_ with a generator key different from the public key used for transaction signatures eventually. This is due to [security benefits](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0044.md#generator-key) added by the concept of generator keys.
+
+#### Invalid Ed25519 Public Key
```

Comment:
@sergeyshemyakov Did you verify this part also for the library that we use in our implementation? If not, I think it would be good if you could do that.

Reply:
I have checked, and the public key should be different. The little endian encoding of 2^255-1 should be `0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe`, note that little endiannes holds bitwise as described in the end of this section: https://datatracker.ietf.org/doc/html/rfc8032#section-2.
I have asked Shu about the implementation library, I will also try to check the actual behavior of SDK.